src/Scene.cpp

```diff
@@ -46,24 +30,78 @@ void Scene::ClearGeometries()
 // Rendering function that draws our scene to the frame buffer
 void Scene::Render() const
 {
+	
 	glClearColor(0.2f, 0.2f, 0.2f, 1.0f);
-	glClear(GL_COLOR_BUFFER_BIT);
-	glUseProgram(program);
-
-	glm::mat4 model = glm::mat4(1.0); // placeholder 
-	glm::mat4 MVP = camera->ProjectionMatrix* camera->ViewMatrix * model;
-
-	glUniformMatrix4fv(glGetUniformLocation(program,"MVP"), 1, GL_FALSE, &MVP[0][0]);
-
-	for(auto geometry : geometries)
+	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+	// Draw non stencil objects here
+	// TODO ...
+	for(int i = 0; i < geometries.size(); i++)
 	{
-		// bind our shader program and the vertex array object containing our
-		// scene geometry, then tell OpenGL to draw our geometry
-		glBindVertexArray(geometry->vertexArray);
-		glDrawArrays(geometry->renderMode, 0, geometry->vertices.size());
+		if(!selected[i])
+		{
+			program->use();
+			glStencilMask(0x00);
+			glm::mat4 model = glm::mat4(1.0); // placeholder
+			glm::mat4 MVP = camera->ProjectionMatrix* camera->ViewMatrix * model;
+			program->setMat4("MVP", MVP);
+			
+			glBindVertexArray(geometries[i]->vertexArray);
+			glDrawArrays(geometries[i]->renderMode, 0, geometries[i]->vertices.size());
+		}
+		else
+		{
+			RenderStencil(geometries[i]);
+		}
 	}
-
+	
+	glEnable(GL_DEPTH_TEST);
+	
 	// reset state to default (no shader or geometry bound)
 	glBindVertexArray(0);
 	glUseProgram(0);
 }
+
+
+void Scene::RenderStencil(Geometry* geometry) const
+{
+	program->use();
+	
+	glStencilFunc(GL_ALWAYS, 1, 0xFF);
+	glStencilMask(0xFF);
+	
+	glm::mat4 model = glm::mat4(1.0); // placeholder
+	glm::mat4 MVP = camera->ProjectionMatrix* camera->ViewMatrix * model;
+	
+	program->setMat4("MVP", MVP);
+	
+		// bind our shader program and the vertex array object containing our
+		// scene geometry, then tell OpenGL to draw our geometry
+	glBindVertexArray(geometry->vertexArray);
+	glDrawArrays(geometry->renderMode, 0, geometry->vertices.size());
+	
+	glStencilFunc(GL_NOTEQUAL, 1, 0xFF);
+	glStencilMask(0x00);
+	glDisable(GL_DEPTH_TEST);
```

Comment:
It is confusing as when we need GL_DEPTH_TEST as it is getting turned on and off in different places.

Is it possible to localize the enabling and disabling of it? Better to have redundant but understandable code. If not, please add brief comments so readers can follow along.

A possible solution is to enable GL_DEPTH_TEST during Scene::Render() and disabling it at the end of the function.

Reply:
remove some unecessary DepthTest should be clear now.